File: tracking/discord_publish.py
"""Discord publish state is the board. VOID does not retract a post.

Matt, 2026-09-23: Discord is the source of truth. The app shows the same bets
the channel still has, and it does not present a bet the channel never got.

A confirmed post is a `push_sent` row of kind `discord_signal` (pre-game) or
`discord_live` (in-play). The row is written only after Discord accepts the
message. VOID does not delete that message and does not delete the row, so
the ledger IS the channel.

`opening_signals` is the CLV shadow track, not the board.
`opening_lock_is_live` / `discord_published_exists_sql` are the join a
reader should use (a capture is live only when this ledger has the key).
Capture, system health, and the publisher do not call them yet — those
readers still treat a capture row as locked. Wiring them is TBD. Nothing
here deletes a capture.
"""

from __future__ import annotations

from tracking.publish_keys import KEY_PARTS, LIVE_KEY_PREFIX

# The two kinds that are a message in a Discord channel. `new_bet` is the
# phone push and stores no message. `discord_free_pick.message_id` stores a
# lock_key, not a snowflake, and is a different channel.
DISCORD_LED_KINDS = ("discord_signal", "discord_live")


def lock_key_for_pick(row: dict, *, live: bool = False) -> str:
    """The same string `lock_key_sql` / `live_lock_key_sql` mint.

    Null and empty identity parts are omitted, matching
    `COALESCE(':' || col, '')`.
    """
    tail = "".join(f":{row[c]}" for c in KEY_PARTS if row.get(c))
    if live:
        return (f"{LIVE_KEY_PREFIX}{row['game_id']}:{row['model_id']}:"
                f"{row.get('pick_side') or ''}{tail}")
    return f"{row['game_id']}:{row['model_id']}{tail}"


def discord_published_exists_sql(lock_expr: str) -> str:
    """SQL: this lock was posted to a Discord channel and the post stands.

    Kinds are the channel posts only. A `new_bet` row is not Discord.
    """
    kinds = ", ".join(f"'{k}'" for k in DISCORD_LED_KINDS)
    return (f"EXISTS (SELECT 1 FROM push_sent s "
            f"WHERE s.lock_key = {lock_expr} AND s.kind IN ({kinds}))")


def discord_led_visible(condition_status: str | None,
                        published: bool | None) -> bool:
    """Whether a BET is an active Discord-led bet.

    `published` True  — the ledger has the lock. VOID does not hide it.
    `published` False — Discord does not have it. Not an active Discord bet.
    `published` None  — the ledger could not be read. A non-VOID bet keeps
    the pre-ledger display so a missing view cannot blank the board. A VOID
    stays hidden: there is no evidence the channel still shows it.
    """
    if published is True:
        return True
    if condition_status == "VOID":
        return False
    if published is False:
        return False
    return True


def void_hidden_from_board(condition_status: str | None,
                           published: bool | None) -> bool:
    """Today's scored list drops a VOID the channel does not still show.

    A published VOID stays: members who saw Discord must see the same bet.
    """
    if condition_status != "VOID":
        return False
    return published is not True


def opening_lock_is_live(captured: bool, discord_published: bool) -> bool:
    """Predicate for a reader that wants capture AND a Discord post.

    Not called by capture, system health, or the publisher yet. Those still
    treat "row exists" as locked. The capture row is not deleted on VOID
    (ON CONFLICT would just re-lock it). Wiring this in is TBD.
    """
    return bool(captured) and bool(discord_published)

File: mobile/src/lib/discordPublish.ts
/**
 * Discord publish state is the board. VOID does not retract a post.
 *
 * Matt, 2026-09-23: Discord is the source of truth. Today / Signals show a
 * bet the channel still has, including one later marked VOID. A bet the
 * channel never got is not an active Discord-led bet.
 *
 * The ledger is `v_discord_published` (push_sent kinds discord_signal and
 * discord_live). The app cannot read push_sent itself. A failed read is
 * `unknown`: non-VOID bets keep the previous display, VOID stays hidden.
 *
 * The lock key is tracking/publish_keys.py. Null and empty identity parts
 * are omitted, matching COALESCE(':' || col, '').
 */

export type DiscordPublish = 'published' | 'unpublished' | 'unknown';

const KEY_PARTS = ['player_id', 'player_key', 'prop_market'] as const;

type LockPick = {
  game_id: string;
  model_id: string;
  pick_side?: string | null;
  player_id?: string | null;
  player_key?: string | null;
  prop_market?: string | null;
  is_live?: boolean | null;
  signal_type?: string | null;
  pick_id?: number;
};

export function lockKeyForPick(p: LockPick): string {
  const tail = KEY_PARTS.map((c) => p[c]).filter((c) => c).join(':');
  if (p.is_live) {
    const base = `live:${p.game_id}:${p.model_id}:${p.pick_side ?? ''}`;
    return tail ? `${base}:${tail}` : base;
  }
  return tail ? `${p.game_id}:${p.model_id}:${tail}` : `${p.game_id}:${p.model_id}`;
}

/**
 * A BET is on the Discord board.
 *
 * published  — the channel has it. VOID does not hide it.
 * unpublished — the channel does not. Not an active Discord-led bet.
 * unknown / omitted — ledger unread. Non-VOID stays; VOID stays hidden.
 */
export function discordLedVisible(
  conditionStatus: string | null | undefined,
  publish: DiscordPublish | undefined,
): boolean {
  if (publish === 'published') return true;
  if (conditionStatus === 'VOID') return false;
  if (publish === 'unpublished') return false;
  return true;
}

/** Today's scored list drops a VOID the channel does not still show. */
export function voidHiddenFromBoard(p: {
  condition_status: string | null;
  discordPublish?: DiscordPublish;
}): boolean {
  return p.condition_status === 'VOID' && p.discordPublish !== 'published';
}

/**
 * Still open for the user to act on — Track and betslip render.
 *
 * `result == null` is an unsettled pick. A VOID writes `result = 'NO_ACTION'`
 * as its marker, not as a settlement, so a VOID the channel still shows is
 * the same open bet (Matt, 2026-09-23) and keeps its actions. Without this a
 * published VOID drew on the board with no Track and no betslip — the
 * 2026-09-26 UFC card. The settled record still excludes it
 * (passesRecordFilter).
 */
export function openForAction(p: {
  result: string | null;
  condition_status: string | null;
  discordPublish?: DiscordPublish;
}): boolean {
  if (p.result == null) return true;
  return p.result === 'NO_ACTION'
    && p.condition_status === 'VOID'
    && p.discordPublish === 'published';
}

const CHUNK = 80;

/**
 * Ledger read for these picks. `null` means the read failed (view missing,
 * permission, network) — callers must treat that as `unknown`, not as
 * "Discord has nothing".
 */
export async function fetchPublishedLockKeys(picks: LockPick[]): Promise<Set<string> | null> {
  const keys = [...new Set(picks.filter((p) => p.signal_type === 'BET').map(lockKeyForPick))];
  if (keys.length === 0) return new Set();
  const { supabase } = await import('./supabase');
  const found = new Set<string>();
  const chunks: string[][] = [];
  for (let i = 0; i < keys.length; i += CHUNK) chunks.push(keys.slice(i, i + CHUNK));
  const results = await Promise.all(chunks.map((chunk) =>
    supabase.from('v_discord_published').select('lock_key').in('lock_key', chunk)));
  for (const { data, error } of results) {
    if (error) return null;
    for (const row of (data ?? []) as { lock_key: string }[]) found.add(row.lock_key);
  }
  return found;
}

export function discordPublishState(key: string, published: Set<string> | null): DiscordPublish {
  if (published == null) return 'unknown';
  return published.has(key) ? 'published' : 'unpublished';
}

export async function attachDiscordPublish<T extends LockPick & {
  signal_type?: string | null;
  discordPublish?: DiscordPublish;
}>(picks: T[]): Promise<T[]> {
  const published = await fetchPublishedLockKeys(picks);
  return picks.map((p) => ({
    ...p,
    discordPublish: p.signal_type === 'BET'
      ? discordPublishState(lockKeyForPick(p), published)
      : p.discordPublish,
  }));
}
